solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs

```diff
@@ -1,26 +1,209 @@
+            }
+        };
+        let channel_id = port_channel.channel_id();
+        let port_id = port_channel.port_id();
+        let seeds =
+            [port_id.as_bytes(), channel_id.as_bytes(), denom.as_bytes()];
+        let (escrow_account_key, _bump) =
+            Pubkey::find_program_address(&seeds, &crate::ID);
+        Ok(escrow_account_key)
+    }
+}
+
+impl TryFrom<&AccountIdx> for Pubkey {
+    type Error = InvalidAccountIdVariant;
+
+    fn try_from(value: &AccountIdx) -> Result<Self, Self::Error> {
+        match value {
+            AccountIdx::Signer(signer) => Ok(signer.0),
+            AccountIdx::Escrow(_) => {
+                Err(InvalidAccountIdVariant::NotSignerAccount)
+            }
+        }
+    }
+}
 
-impl TokenTransferExecutionContext for IbcStorage<'_, '_> {
+impl TokenTransferExecutionContext for IbcStorage<'_, '_, '_> {
     fn send_coins_execute(
         &mut self,
-        _from: &Self::AccountId,
-        _to: &Self::AccountId,
-        _amt: &PrefixedCoin,
+        from: &Self::AccountId,
+        to: &Self::AccountId,
+        amt: &PrefixedCoin,
     ) -> Result<(), TokenTransferError> {
-        //let sender_id = from.to_string();
-        //let receiver_id = to.to_string();
-        //let base_denom = amt.denom.base_denom.to_string();
-        todo!()
+        msg!(
+            "Sending coins from account {} to account {}, trace path {}, base \
+             denom {}",
+            from,
+            to,
+            amt.denom.trace_path,
+            amt.denom.base_denom
+        );
+        let base_denom = amt.denom.base_denom.to_string();
+        let sender_id = Pubkey::try_from(from).unwrap_or_else(|_| {
+            from.get_escrow_account(base_denom.clone()).unwrap()
+        });
+        let receiver_id = Pubkey::try_from(to).unwrap_or_else(|_| {
+            to.get_escrow_account(base_denom.clone()).unwrap()
+        });
+
+        let amount = amt.amount;
+        let amount_in_u64 = check_amount_overflow(amount)?;
+
+        let (_token_mint_key, _bump) =
+            Pubkey::find_program_address(&[base_denom.as_ref()], &crate::ID);
+        let store = self.borrow();
+        let accounts = &store.accounts;
+
+        let sender = get_account_info_from_key(accounts, sender_id)?;
+        let receiver = get_account_info_from_key(accounts, receiver_id)?;
+        let token_program = get_account_info_from_key(accounts, spl_token::ID)?;
+
+        if matches!(from, AccountIdx::Escrow(_)) {
+            let (mint_authority_key, mint_authority_bump) =
+                Pubkey::find_program_address(&[MINT_ESCROW_SEED], &crate::ID);
+
+            let mint_authority =
+                get_account_info_from_key(accounts, mint_authority_key)?;
+
+            let bump_vector = mint_authority_bump.to_le_bytes();
+            let inner = vec![MINT_ESCROW_SEED, bump_vector.as_ref()];
+            let outer = vec![inner.as_slice()];
```

Comment:
No need for vectors:

```suggestion
            let seeds = [MINT_ESCROW_SEED, bump_vector.as_ref()];
            let seeds = seeds.as_ref();
            let seeds = core::slice::from_ref(&seeds);
```

Same in the else branch.